hir/src/ir/traits/types.rs

```diff
@@ -68,12 +68,67 @@ derive! {
 }
 
 derive! {
-    /// Op expects all operands and results to be of the same type
+    /// Op expects all operands and results to be of the same type.
+    /// NOTE: Operations that implement this trait must also explicitly implement
+    /// [`SameTypeOperands`]. This can be achieved by using the "traits" field in the [#operation]
+    /// macro.
+    /// Like so:
     ///
-    /// TODO(pauls): Implement verification for this. Ideally we could require `SameTypeOperands`
-    /// as a super trait, check the operands using its implementation, and then check the results
-    /// separately
-    pub trait SameOperandsAndResultType {}
+    /// #[operation (
+    ///         dialect = ArithDialect,
+    ///         traits(UnaryOp, SameTypeOperands, SameOperandsAndResultType),
+    ///         implements(InferTypeOpInterface, MemoryEffectOpInterface)
+    /// )]
+    /// pub struct SomeOp {
+    ///       (...)
+    /// }
+    pub trait SameOperandsAndResultType: SameTypeOperands {}
+
+    verify {
+        fn operands_and_result_are_the_same_type(op: &Operation, context: &Context) -> Result<(), Report> {
+            let mut operands = op.operands().iter();
+            if let Some(first_operand) = operands.next() {
```

Comment:
I think we should panic if the operation has no operands but derives `SameTypeOperands` (and by extension `SameOperandsAndResultType`), since the trait is meaningless for an operation with no operands. I would prefer to panic in that situation vs raise a compiler diagnostic, since it represents an internal compiler error that can't meaningfully be fixed without modifying the operation definition.